lib/code-caller-docker.js

```diff
@@ -0,0 +1,742 @@
+ * already been pulled before PrairieLearn starts up.
+ */
+async function ensureImage() {
+  const imageName = getExecutorImageName();
+  const image = docker.getImage(imageName);
+  try {
+    await image.inspect();
+  } catch (e) {
+    if (e.statusCode === 404) {
+      const dockerAuth = await dockerUtil.setupDockerAuthAsync();
+      await docker.createImage(dockerAuth, { fromImage: imageName });
+    } else {
+      throw e;
+    }
+  }
+}
+
+/** @typedef {import('./code-caller-shared').CodeCaller} CodeCaller */
+/** @typedef {import('./code-caller-shared').CallType} CallType */
+
+/**
+ * @implements {CodeCaller}
+ */
+class DockerCaller {
+  constructor(options = { questionTimeoutMilliseconds: 5000 }) {
+    /** @type {CallerState} */
+    this.state = CREATED;
+    this.uuid = uuidv4();
+
+    this.debug('enter constructor()');
+
+    /** @type {import('dockerode').Container} */
+    this.container = null;
+    this.callback = null;
+    this.timeoutID = null;
+    this.callCount = 0;
+    this.ensureChildMutex = new Mutex();
+    this.ensureDirectoryMutex = new Mutex();
+
+    this.options = options;
+
+    // These will accumulate output from the container.
+    this.outputStdout = '';
+    this.outputStderr = '';
+
+    // for error logging
+    this.lastCallData = null;
+
+    this.coursePath = null;
+
+    this._checkState();
+
+    this.debug(`exit constructor(), state: ${String(this.state)}, uuid: ${this.uuid}`);
+  }
+
+  /**
+   * Wrapper around `debug` that automatically includes UUID and the caller state.
+   *
+   * @param {string} message
+   */
+  debug(message) {
+    const paddedState = this.state.toString().padEnd(15);
+    debug(`[${this.uuid} ${paddedState}] ${message}`);
+  }
+
+  /**
+   * Wrapper around `createBindMount` that includes instance-specific logs.
+   *
+   * @param {string} directory
+   * @param {string} mountpoint
+   */
+  async createBindMount(directory, mountpoint) {
+    this.debug(`creating bind mount for ${directory} at ${mountpoint}`);
+    await instrumented('createBindMount', async (span) => {
+      span.setAttribute('mountpoint', mountpoint);
+      await bindMount.mount(directory, mountpoint);
+    });
+    this.debug(`created bind mount for ${directory} at ${mountpoint}`);
+  }
+
+  /**
+   * Wrapper around `removeBindMount` that includes instance-specific logs.
+   *
+   * @param {string} mountpoint
+   */
+  async removeBindMount(mountpoint) {
+    this.debug(`removing bind mount at ${mountpoint}`);
+    await instrumented('removeBindMount', async (span) => {
+      span.setAttribute('mountpoint', mountpoint);
+      await bindMount.umount(mountpoint);
+    });
+    this.debug(`removed bind mount at ${mountpoint}`);
+  }
+
+  /**
+   * Allows this caller to prepare for execution of code from a particular
+   * course.
+   *
+   * @param {string} coursePath
+   */
+  async prepareForCourse(coursePath) {
```

Comment:
If we call `ensureChild()` as the first action in `prepareForCourse()` then we can remove the directory mutex.

Reply:
Done in fab8b2d!